src/main/kotlin/org/radarcns/gateway/kafka/AvroProcessor.kt

```diff
@@ -0,0 +1,161 @@
+            throw IllegalArgumentException("Records should be an array")
+        }
+
+        records.forEach { record ->
+            if (isNullField(record["key"])) {
+                throw IllegalArgumentException("Missing key field in record")
+            }
+            if (isNullField(record["value"])) {
+                throw IllegalArgumentException("Missing value field in record")
+            }
+            processKey(record["key"], auth)
+        }
+    }
+
+    /** Parse single record key.  */
+    @Throws(IOException::class, NotAuthorizedException::class)
+    private fun processKey(key: JsonNode, auth: Auth) {
+        if (!key.isObject) {
+            throw IllegalArgumentException("Field key must be a JSON object")
+        }
+
+        val project = key["projectId"]
+        if (project != null) {
+            if (project.isNull) {
+                // no project ID was provided, fill it in for the sender
+                val newProject = Json.mapper.createObjectNode()
+                newProject.put("string", auth.projectIds[0])
+                (key as ObjectNode).set("projectId", newProject)
+            } else {
+                // project ID was provided, it should match one of the validated project IDs.
+                val projectId = project["string"]?.asText() ?:
+                        throw IllegalArgumentException(
+                                "Project ID should be wrapped in string union type")
+
+                if (!auth.projectIds.contains(projectId)) {
+                    throw NotAuthorizedException("record projectId '$projectId' does not match"
+                           + " authenticated user ID '${auth.userId}'")
+                }
+            }
+        }
+
+        val user = key["userId"]
+        if (user != null && user.asText() != auth.userId) {
+            throw NotAuthorizedException(
+                    "record userId '${user.asText()}' does not match authenticated user ID "
+                            + "'${auth.userId}'")
+        }
+        val source = key["sourceId"]
+        if (source != null && !auth.sourceIds.contains(source.asText())) {
+            throw NotAuthorizedException(
+                    "record sourceId '${source.asText()}' has not been added to JWT allowed "
+                            + "IDs ${auth.sourceIds}.")
+        }
+    }
+
+    private class Auth(jwt: DecodedJWT) {
+        val projectIds: List<String>
+        val userId: String = jwt.subject
+        val sourceIds: Collection<String>
+        init {
+            // get roles as a list, empty if not set
+            val rolesClaim = jwt.getClaim(ROLES_CLAIM)?.asList(String::class.java)
+                    ?: Collections.emptyList()
+
+            projectIds = rolesClaim.stream()
+                    .filter { it.endsWith(PARTICIPANT_SUFFIX) }
+                    .map { it.substring(0, it.length - PARTICIPANT_SUFFIX.length) }
+                    .collect(Collectors.toList())
```

Comment:
I think we should check for the `MEASUREMENT_CREATE` permission, rather than checking for the participant role. I.e. extract all projects from the roles claim, then filter them using [`RadarAuthorization.checkPermissionOnProject()`](https://radar-cns.github.io/ManagementPortal/ra-javadoc/org/radarcns/auth/authorization/RadarAuthorization.html#checkPermissionOnProject-com.auth0.jwt.interfaces.DecodedJWT-org.radarcns.auth.authorization.Permission-java.lang.String-). This keeps the responsibility of linking permissions to roles central to `radar-auth`.

Reply:
[ManagementPortalAuthenticationFilter](https://github.com/RADAR-CNS/RADAR-Gateway/blob/release-0.1.3/src/main/kotlin/org/radarcns/gateway/filter/ManagementPortalAuthenticationFilter.kt#L41) checks the `MEASUREMENT_CREATE` permission in  because it can be checked without inspecting the content. The problem is that without parsing the request content, it is not known what projects permissions are needed for.

Reply:
And if a project ID is not set in the content, we discussed before that the Gateway should fill in that information. So the Avro filter needs to find out what project to fill in. Alternatively, there could be a `radar-auth` function to parse projects where the users has a certain role? Like
```java
Collection<String> projectIds = RadarAuthorization.getAuthorizedProjects(jwt, ROLE_PARTICIPANT);
```

Reply:
I'd rather not rely on a `try-catch` construct for a simple `if` condition. Could there be a boolean `hasPermissionOnProject`?

Reply:
Yep that makes a lot of sense to me, i'll get on it right away!